File: Medium/39MergeOverlappingIntervals.js
function mergeOverlappingIntervals(array) {
  // Write your code here.
  array.sort((a, b) => a[0] - b[0]);

  let mergedIntervals = [];
  let currentIntervals = array[0];
  mergedIntervals.push(currentIntervals);
  for (let nextIntervals of array) {
    let endCurrentInterval = currentIntervals[1];
    let startNextInterval = nextIntervals[0];
    let endNextInterval = nextIntervals[1];

    if (endCurrentInterval >= startNextInterval) {
      currentIntervals[1] = Math.max(endCurrentInterval, endNextInterval);
    } else {
      currentIntervals = nextIntervals;
      mergedIntervals.push(currentIntervals);
    }
  }
  return mergedIntervals;
}

console.log("====================================");
console.log(
  mergeOverlappingIntervals([
    [1, 2],
    [4, 7],
    [3, 5],
    [9, 10],
    [6, 8],
  ])
);
console.log("====================================");

File: Medium/39MergeOverlappingIntervals.py
def mergeOverlappingIntervals(intervals):
    # Write your code here.
    sortedIntervals = sorted(intervals, key=lambda x: x[0])

    mergedIntervals = []
    currentInterval = sortedIntervals[0]
    mergedIntervals.append(currentInterval)

    for nextInterval in sortedIntervals:
        _, currentIntervalEnd = currentInterval
        nextIntervalStart, nextIntervalEnd = nextInterval

        if currentIntervalEnd >= nextIntervalStart:
            currentInterval[1] = max(currentIntervalEnd, nextIntervalEnd)
        else:
            currentInterval = nextInterval
            mergedIntervals.append(currentInterval)

    return mergedIntervals


print(mergeOverlappingIntervals([
    [1, 2],
    [3, 5],
    [4, 7],
    [6, 8],
    [9, 10]
]))
